catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java

```diff
@@ -14,51 +14,77 @@
+import java.util.stream.Collectors;
 
 import javax.xml.bind.JAXBElement;
+import javax.xml.bind.Marshaller;
 
 import org.codice.ddf.parser.Parser;
 import org.codice.ddf.parser.ParserConfigurator;
 import org.codice.ddf.parser.ParserException;
 
 import com.google.common.base.Charsets;
-import com.google.common.io.CharStreams;
-import com.google.common.io.FileBackedOutputStream;
 
 import ddf.catalog.data.Metacard;
 import ddf.catalog.data.impl.AttributeImpl;
+import ddf.catalog.federation.FederationException;
 import ddf.catalog.operation.CreateRequest;
+import ddf.catalog.operation.CreateResponse;
 import ddf.catalog.operation.DeleteRequest;
+import ddf.catalog.operation.DeleteResponse;
 import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.operation.UpdateResponse;
 import ddf.catalog.plugin.PluginExecutionException;
+import ddf.catalog.plugin.PostIngestPlugin;
 import ddf.catalog.plugin.PreIngestPlugin;
 import ddf.catalog.plugin.StopProcessingException;
-import ddf.catalog.registry.api.metacard.RegistryObjectMetacardType;
 import ddf.catalog.registry.common.RegistryConstants;
+import ddf.catalog.registry.common.metacard.RegistryObjectMetacardType;
+import ddf.catalog.registry.federationadmin.service.FederationAdminService;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+import ddf.catalog.util.impl.Requests;
 import oasis.names.tc.ebxml_regrep.xsd.rim._3.ExternalIdentifierType;
 import oasis.names.tc.ebxml_regrep.xsd.rim._3.RegistryObjectType;
 
-public class IdentificationPlugin implements PreIngestPlugin {
+/**
+ * IdentificationPlugin is a Pre/PostIngestPlugin that assigns a localID when a metacard is added to the
+ * catalog and an originID to a registry metacard during creation. It also ensures that duplicate
+ * registry-ids are not added to the catalog.
+ */
+public class IdentificationPlugin implements PreIngestPlugin, PostIngestPlugin {
 
     private Parser parser;
 
-    private ParserConfigurator configurator;
+    private FederationAdminService federationAdmin;
+
+    private ParserConfigurator marshalConfigurator;
+
+    private ParserConfigurator unmarshalConfigurator;
+
+    private Set<String> registryIds = ConcurrentHashMap.newKeySet();
 
     @Override
     public CreateRequest process(CreateRequest input)
             throws PluginExecutionException, StopProcessingException {
-        List<Metacard> metacards = input.getMetacards();
-
-        for (Metacard metacard : metacards) {
-            Set<String> metacardTags = metacard.getTags();
-            if (metacardTags.contains(RegistryConstants.REGISTRY_TAG)) {
-                setMetacardExtID(metacard);
+        if (Requests.isLocal(input)) {
+            for (Metacard metacard : input.getMetacards()) {
+                if (metacard.getTags()
+                        .contains(RegistryConstants.REGISTRY_TAG)) {
+                    if (registryIds.contains(getRegistryId(metacard))) {
```

Comment:
We could do the id check before we do any processing to be more efficient. We have another ticket to do some work in this plugin so we might change it then. Right now this code will experience very low throughput so it isn't a big concern but we would like to be efficient as possible.